wpcom-vip-two-factor/is-jetpack-sso.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace Automattic\VIP\TwoFactor;
+
+// muplugins_loaded fires before cookie constants are set
+wp_cookie_constants();
+
+define( 'VIP_IS_JETPACK_SSO_COOKIE', AUTH_COOKIE . '_vip_jetpack_sso' );
+define( 'VIP_IS_JETPACK_SSO_2SA_COOKIE', AUTH_COOKIE . '_vip_jetpack_sso_2sa' );
+
+add_action( 'jetpack_sso_handle_login', function( $user, $user_data ) {
+	add_action( 'set_auth_cookie', function( $auth_cookie, $expire, $expiration, $user_id, $scheme, $token ) use ( $user_data ) {
+		$secure = is_ssl();
+
+		$sso_cookie = create_cookie( $user_id, $expire, VIP_IS_JETPACK_SSO_COOKIE );
+		setcookie( VIP_IS_JETPACK_SSO_COOKIE, $sso_cookie, $expire, COOKIEPATH, COOKIE_DOMAIN, $secure, true );
+
+		if ( $user_data->two_step_enabled ) {
+			$sso_2sa_cookie = create_cookie( $user_id, $expire, VIP_IS_JETPACK_SSO_2SA_COOKIE );
+			setcookie( VIP_IS_JETPACK_SSO_2SA_COOKIE, $sso_2sa_cookie, $expire, COOKIEPATH, COOKIE_DOMAIN, $secure, true );
+		}
+	}, 10, 6 );
+}, 10, 2 );
+
+add_action( 'clear_auth_cookie', function() {
+	setcookie( VIP_IS_JETPACK_SSO_COOKIE, ' ', time() - YEAR_IN_SECONDS, COOKIEPATH, COOKIE_DOMAIN );
+	setcookie( VIP_IS_JETPACK_SSO_2SA_COOKIE, ' ', time() - YEAR_IN_SECONDS, COOKIEPATH, COOKIE_DOMAIN );
+} );
+
+function is_jetpack_sso() {
+	if ( ! is_user_logged_in() ) {
+		return false;
+	}
+
+	if ( ! isset( $_COOKIE[ VIP_IS_JETPACK_SSO_COOKIE ] ) ) {
+		return false;
+	}
+
+	$cookie = $_COOKIE[ VIP_IS_JETPACK_SSO_COOKIE ];
+	return verify_cookie( $cookie, VIP_IS_JETPACK_SSO_COOKIE );
+}
+
+function is_jetpack_sso_two_step() {
+	if ( ! is_jetpack_sso() ) {
+		return false;
+	}
+
+	if ( ! isset( $_COOKIE[ VIP_IS_JETPACK_SSO_2SA_COOKIE ] ) ) {
+		return false;
+	}
+
+	$cookie = $_COOKIE[ VIP_IS_JETPACK_SSO_2SA_COOKIE ];
+	return verify_cookie( $cookie, VIP_IS_JETPACK_SSO_2SA_COOKIE );
+}
+
+function create_cookie( $user_id, $expiration, $scheme ) {
+	$user = get_userdata( $user_id );
+
+	if ( ! $user ) {
+		// Invalid user
+		return;
+	}
+
+	$key = wp_hash( $user->user_login . '|' . $expiration, $scheme );
+	$hash = hash_hmac( 'md5', $user->user_login . '|' . $expiration, $key );
```

Comment:
I think we need to include a unique nonce in the cookie value (and hashed message) to make it more unique and tamper-resistant. Otherwise, the only thing that's unique about the key for a given user is the username and expiration, which could potentially make it easier to attack because the same key will be reused across every session.

Core includes both a fragment of the password (hashed hopefully) and the session token in the key in `wp_generate_auth_cookie()` - for this, i think just a nonce from `wp_create_nonce( 'vip_jetpack_sso' )` could be good enough?

Reply:
>  for this, i think just a nonce from wp_create_nonce( 'vip_jetpack_sso' ) could be good enough?

I think we need something that won't expire (at least not before the cookie expires).

Reply:
It can be anything and doesn't actually need to be secret - it'll be stored in the cookie in plain text (along the user id + expiration) so that we can recreate the key again, so even if the nonce technically expired, we're not doing anything with nonce verification, just using it for some entropy. We can use `wp_generate_password()` also.

Reply:
What we're after is the same thing core does for auth cookies...

```
$cookie = $user->user_login . '|' . $expiration . '|' . $token . '|' . $hash;
```